test/rubygems/test_gem_package_tar_reader_entry.rb

```diff
@@ -185,6 +185,13 @@ def test_readpartial
     assert_equal @contents[0...100], @entry.readpartial(100)
   end
 
+  def test_readpartial_outbuf
+    buffer = String.new(capacity: 100, encoding: Encoding::BINARY)
+    partial = @entry.readpartial(100, buffer)
+    assert_equal @contents[0...100], partial
+    assert_same buffer, partial
```

Comment:
Good test. I wonder if it's worth doing one using the gzip setup lower in this test file. I'm not sure it will catch the failure case though, since I think you need a few files so that it can overread one of them.